File: streamlit-app/src/blueprint_pipeline.py
"""
Blueprint Pipeline — 4-step deterministic QE blueprint generation using Claude.
"""
import json
from typing import Optional, Callable


class BlueprintPipeline:
    """Generates Quiet Eye conditioning blueprints via a 4-step Claude pipeline."""

    # ── Core Principles (injected into every prompt) ─────────
    QE_PRINCIPLES = """
QUIET EYE SCIENCE (Joan Vickers):
- The Quiet Eye is the final fixation on a specific target before a critical motor action
- Elite performers hold this fixation 62% longer than non-elite
- DAN (Dorsal Attention Network) = deliberate, planned eye movements
- VAN (Ventral Attention Network) = involuntary attention capture (distraction)
- Our goal: train DAN dominance through repetitive gaze conditioning

FLOW STATE AWARENESS (200ms Lead):
- EYES lock on the CURRENT physical target (foveal, DAN-controlled)
- AWARE holds the NEXT target in peripheral vision (200ms preparation lead)
- This creates seamless gaze transitions without attention gaps

THE FOUR CUES (every corner, same language, same sequence):
1. "Eyes Braking Marker — Aware Apex"
2. "Eyes Apex — Aware Exit"
3. "Eyes Exit — Aware Straight"
4. "Eyes Straight — Aware Braking Marker"
"""

    def __init__(self, api_engine):
        self.api = api_engine

    def generate_blueprint(self, track_config, progress_cb=None):
        """
        Run the full 4-step pipeline for all corners.

        Args:
            track_config: dict with trackName, vehicleType, skillLevel, corners[]
            progress_cb: callback(percent, message)

        Returns: complete blueprint dict
        """
        corners = track_config.get('corners', [])
        template = self._get_template(track_config.get('vehicleType', 'car'))
        total_steps = len(corners) * 3 + 1  # 3 per-corner steps + 1 protocol assembly
        step_count = 0

        processed_corners = []

        for i, corner in enumerate(corners):
            corner_name = corner.get('name', f'Corner {i+1}')

            # Step 1: Classify
            if progress_cb:
                pct = int((step_count / total_steps) * 90)
                progress_cb(pct, f"Step 1/4: Classifying {corner_name}...")

            classified = self._step1_classify(corner, template, track_config)
            step_count += 1

            # Step 2: Gaze Sequence
            if progress_cb:
                pct = int((step_count / total_steps) * 90)
                progress_cb(pct, f"Step 2/4: Building gaze sequence for {corner_name}...")

            gaze_data = self._step2_gaze_sequence(classified, template, track_config)
            step_count += 1

            # Step 3: Risk Analysis
            if progress_cb:
                pct = int((step_count / total_steps) * 90)
                progress_cb(pct, f"Step 3/4: Risk analysis for {corner_name}...")

            risk_data = self._step3_risk_analysis(classified, gaze_data, template, track_config)
            step_count += 1

            # Merge all data for this corner
            processed = {**classified, **gaze_data, **risk_data}
            processed_corners.append(processed)

        # Step 4: Protocol Assembly (all corners at once)
        if progress_cb:
            progress_cb(85, "Step 4/4: Assembling training protocol...")

        protocol = self._step4_protocol(processed_corners, template, track_config)

        # Build final blueprint
        blueprint = {
            'trackName': track_config.get('trackName', 'Unknown Track'),
            'vehicleType': track_config.get('vehicleType', 'car'),
            'skillLevel': track_config.get('skillLevel', 'intermediate'),
            'clientName': track_config.get('clientName', ''),
            'cornerCount': len(processed_corners),
            'corners': processed_corners,
            'protocol': protocol,
            'metadata': {
                'generator': 'Lap Blueprint Generator (Streamlit)',
                'model': self.api.claude_model,
                'pipeline_version': '2.0'
            }
        }

        if progress_cb:
            progress_cb(100, "Blueprint complete!")

        return blueprint

    # ── Step 1: Corner Classification ────────────────────────

    def _step1_classify(self, corner, template, track_config):
        system = f"""You are a Quiet Eye corner classification specialist for motorsport.
{self.QE_PRINCIPLES}

VEHICLE: {template['vehicleType']}
{template.get('systemPromptSection', '')}"""

        user = f"""Classify this corner for Quiet Eye conditioning:

TRACK: {track_config.get('trackName', 'Unknown')}
SKILL LEVEL: {track_config.get('skillLevel', 'intermediate')}

CORNER DATA:
{json.dumps(corner, indent=2, default=str)}

Return ONLY valid JSON:
{{
  "cornerNumber": <int>,
  "name": "<corner name>",
  "type": "hairpin|tight|sweeper|kink|chicane|esses|offcamber|medium",
  "direction": "left|right|unknown",
  "severity": "very_tight|tight|medium|fast|flat_out",
  "qeDifficulty": "trivial|simple|moderate|challenging|expert",
  "primaryVanTrigger": "target_fixation|early_apex_look|peripheral_threat|instrument_glance|unexpected_threat",
  "gazeDominanceRecommendation": "early_commitment|extended_fixation|rapid_transition",
  "classificationReasoning": "<1-2 sentences explaining the QE challenge>"
}}"""

        return self._call_with_retry(system, user)

    # ── Step 2: Gaze Sequence ────────────────────────────────

    def _step2_gaze_sequence(self, classified, template, track_config):
        system = f"""You are a Quiet Eye gaze protocol specialist.
{self.QE_PRINCIPLES}

VEHICLE: {template['vehicleType']}
{template.get('systemPromptSection', '')}

CRITICAL RULES:
- "eyes" must be a SPECIFIC physical feature: kerb color, board number, landmark, cone, barrier
- "aware" must include VEHICLE-SPECIFIC peripheral cue + next visual target
- NEVER abstract: say "red/white inside kerb" not "the apex"
- NEVER mention driving technique — only WHAT TO LOOK AT
- Each "eyes" instruction = ONE clear place for brain to settle
- "cueLabel" follows the pattern: "Eyes [target] — Aware [next target]"
"""

        gaze_phases = template.get('gazePhases', {})
        phases_info = json.dumps(gaze_phases, indent=2) if gaze_phases else "Use standard 5-phase model"

        user = f"""Generate the gaze sequence for this classified corner:

CORNER: {classified.get('name', 'Unknown')} (#{classified.get('cornerNumber', '?')})
TYPE: {classified.get('type', 'medium')} — {classified.get('direction', 'unknown')}
QE DIFFICULTY: {classified.get('qeDifficulty', 'moderate')}
TRACK: {track_config.get('trackName', 'Unknown')}

VEHICLE GAZE PHASES:
{phases_info}

Return ONLY valid JSON:
{{
  "cornerNumber": {classified.get('cornerNumber', 1)},
  "gazeSequence": {{
    "brakeMarkerVisible": {{
      "eyes": "<specific physical target the fovea locks onto>",
      "aware": "<peripheral target + vehicle cue>",
      "fixationDurationSeconds": <0.8-2.0>,
      "cueLabel": "Eyes [target] — Aware [next target]"
    }},
    "brake": {{
      "eyes": "<specific physical target>",
      "aware": "<peripheral target + vehicle cue>",
      "fixationDurationSeconds": <0.5-1.5>,
      "cueLabel": "Eyes [target] — Aware [next target]"
    }},
    "apex": {{
      "eyes": "<specific physical target>",
      "aware": "<peripheral target + vehicle cue>",
      "fixationDurationSeconds": <0.8-2.5>,
      "cueLabel": "Eyes [target] — Aware [next target]"
    }},
    "exit": {{
      "eyes": "<specific physical target>",
      "aware": "<peripheral target + vehicle cue>",
      "fixationDurationSeconds": <0.5-1.5>,
      "cueLabel": "Eyes [target] — Aware [next target]"
    }},
    "nextMarker": {{
      "eyes": "<specific physical target>",
      "aware": "<peripheral target + vehicle cue>",
      "fixationDurationSeconds": <0.5-1.0>,
      "cueLabel": "Eyes [target] — Aware [next target]"
    }}
  }}
}}"""

        return self._call_with_retry(system, user)

    # ── Step 3: Risk Analysis ────────────────────────────────

    def _step3_risk_analysis(self, classified, gaze_data, template, track_config):
        system = f"""You are a motorsport Quiet Eye risk analyst.
{self.QE_PRINCIPLES}

VEHICLE: {template['vehicleType']}

THE FOUR CUES reference:
1. "Eyes Braking Marker — Aware Apex"
2. "Eyes Apex — Aware Exit"
3. "Eyes Exit — Aware Straight"
4. "Eyes Straight — Aware Braking Marker"

A corner is WEAK if:
- QE difficulty is "challenging" or "expert"
- It has 3+ risk factors
- The gaze sequence requires rapid transitions with low confidence targets
"""

        user = f"""Analyze gaze risks for this corner:

CORNER: {classified.get('name', 'Unknown')} (#{classified.get('cornerNumber', '?')})
TYPE: {classified.get('type', 'medium')}
VAN TRIGGER: {classified.get('primaryVanTrigger', 'unknown')}
QE DIFFICULTY: {classified.get('qeDifficulty', 'moderate')}

GAZE SEQUENCE:
{json.dumps(gaze_data.get('gazeSequence', {}), indent=2)}

Return ONLY valid JSON:
{{
  "cornerNumber": {classified.get('cornerNumber', 1)},
  "riskFactors": ["<risk 1>", "<risk 2>", "<risk 3 if applicable>"],
  "isWeakCorner": <true|false>,
  "weakCornerReason": "<reason or empty string>",
  "quietEyeCue": "Settle your eyes on [specific target] — let the [vehicle] follow.",
  "coachingNotes": "<2-3 sentences about gaze targets and awareness ONLY — no driving technique>"
}}"""

        return self._call_with_retry(system, user)

    # ── Step 4: Protocol Assembly ────────────────────────────

    def _step4_protocol(self, processed_corners, template, track_config):
        system = f"""You are the Quiet Eye training protocol architect.
{self.QE_PRINCIPLES}

Design a 5-lap progressive conditioning strategy.
The training follows a 4-week progression:
- Week 1: 0.5x speed, 5-second pauses at each gaze point, WATCH + LISTEN
- Week 2: 0.75x speed, 4-second pauses, LOOK AND CALL (say cues aloud)
- Week 3: 1.0x speed, 3-second pauses, AUDIO CUES ONLY (eyes closed optional)
- Week 4: 1.25x speed, 2-second pauses, AUTOMATIC (full speed, markers only)

Every corner uses the SAME four cue structure:
1. "Eyes Braking Marker — Aware Apex"
2. "Eyes Apex — Aware Exit"
3. "Eyes Exit — Aware Straight"
4. "Eyes Straight — Aware Braking Marker"
"""

        user = f"""Assemble the training protocol for all {len(processed_corners)} corners:

TRACK: {track_config.get('trackName', 'Unknown')}
VEHICLE: {template['vehicleType']}
SKILL LEVEL: {track_config.get('skillLevel', 'intermediate')}

PROCESSED CORNERS:
{json.dumps(processed_corners, indent=2, default=str)}

Return ONLY valid JSON:
{{
  "overallStrategy": "<gaze philosophy focusing on 4-cue rhythm and flow state>",
  "keyPrinciple": "<THE single most important insight for this track>",
  "corners": [
    {{
      "number": <int>,
      "lookAndCall": [
        "Eyes Braking Marker — Aware Apex",
        "Eyes Apex — Aware Exit",
        "Eyes Exit — Aware Straight",
        "Eyes Straight — Aware Braking Marker"
      ],
      "headRotationCue": "Begin head rotation [time] before turn-in...",
      "speedRamp": "25%|50%|100%"
    }}
  ],
  "trainingProtocol": {{
    "dailyMinutes": 15,
    "steps": [
      {{"title": "Watch + Listen (Week 1)", "instruction": "0.5x speed, 5s pauses", "duration": "5min"}},
      {{"title": "Look and Call (Week 2)", "instruction": "0.75x speed, say cues aloud", "duration": "5min"}},
      {{"title": "Audio Only (Week 3)", "instruction": "1.0x speed, no overlays", "duration": "3min"}},
      {{"title": "Full Speed (Week 4)", "instruction": "1.25x speed, automatic", "duration": "2min"}},
      {{"title": "Weak Corner Repetition", "instruction": "3x reps on weak corners", "duration": "5min"}}
    ],
    "weakCornerDrills": "<If Quiet Eye breaks on weak corners, describe remediation>"
  }}
}}"""

        return self._call_with_retry(system, user)

    # ── Helpers ──────────────────────────────────────────────

    def _call_with_retry(self, system, user, max_retries=1):
        """Call Claude pipeline with retry on parse failure."""
        for attempt in range(max_retries + 1):
            try:
                return self.api.call_claude_pipeline(system, user)
            except json.JSONDecodeError as e:
                if attempt < max_retries:
                    user += f"\n\nPREVIOUS RESPONSE FAILED JSON PARSING: {e}\nPlease return ONLY valid JSON, no markdown."
                else:
                    raise
            except Exception:
                if attempt < max_retries:
                    continue
                raise

    def _get_template(self, vehicle_type):
        """Get vehicle-specific template."""
        from src.blueprint_templates import TEMPLATES
        return TEMPLATES.get(vehicle_type, TEMPLATES.get('car', {'vehicleType': vehicle_type}))

File: blueprint-pipeline.js
/* ============================================================
   BLUEPRINT PIPELINE — 4-Step Deterministic Blueprint Engine
   ============================================================
   Generates QE blueprints through 4 focused API calls, each
   validated before proceeding. Temperature 0 for exact
   reproducibility. Vehicle template injected at every step.

   Steps:
   1. CLASSIFY — corner type + QE difficulty + VAN trigger
   2. GAZE SEQUENCE — Eyes/Aware for brake→apex→exit
   3. RISK ANALYSIS — VAN triggers + quietEyeCue + coaching
   4. PROTOCOL ASSEMBLY — strategy, lookAndCall, head rotation

   Steps 1–3 run per corner. Step 4 runs once for the lap.
   ============================================================ */

const BlueprintPipeline = {

    /**
     * Main entry point — generates a full blueprint through the 4-step pipeline.
     * @param {object} trackConfig — { trackName, clientName, vehicleType, skillLevel, trackNotes, corners }
     * @param {function} onProgress — (percent, message) callback
     * @returns {object} — complete blueprint JSON
     */
    async generateBlueprint(trackConfig, onProgress) {
        // 1. Load vehicle template
        const template = BlueprintTemplates.getTemplate(trackConfig.vehicleType);
        const templateId = BlueprintTemplates.getTemplateId(trackConfig.vehicleType);
        const totalCorners = trackConfig.corners.length;

        if (onProgress) onProgress(2, `Loaded ${template.name} template`);

        // 2. Process each corner through Steps 1–3
        const processedCorners = [];
        for (let i = 0; i < totalCorners; i++) {
            const corner = trackConfig.corners[i];
            const cornerLabel = corner.name || `Turn ${i + 1}`;
            const basePercent = 5 + Math.floor((i / totalCorners) * 70);

            // Step 1: Classify
            if (onProgress) onProgress(basePercent, `Step 1/4 — Classifying ${cornerLabel}...`);
            const classified = await this._step1_classify(corner, template, trackConfig);

            // Step 2: Gaze Sequence
            if (onProgress) onProgress(basePercent + 8, `Step 2/4 — Gaze sequence for ${cornerLabel}...`);
            const gazeData = await this._step2_gazeSequence(classified, template, trackConfig);

            // Step 3: Risk Analysis
            if (onProgress) onProgress(basePercent + 16, `Step 3/4 — Risk analysis for ${cornerLabel}...`);
            const riskData = await this._step3_riskAnalysis(classified, gazeData, template, trackConfig);

            // Merge all step outputs into one corner object
            processedCorners.push({
                number: corner.number || (i + 1),
                name: classified.name || cornerLabel,
                type: classified.type || corner.type || 'medium',
                direction: corner.direction || classified.direction || 'unknown',
                severity: classified.severity || corner.severity || 'medium',
                qeDifficulty: classified.qeDifficulty,
                primaryVanTrigger: classified.primaryVanTrigger,
                gazeSequence: gazeData.gazeSequence,
                riskFactors: riskData.riskFactors,
                quietEyeCue: riskData.quietEyeCue,
                coachingNotes: riskData.coachingNotes,
                // Pass through existing data
                approach: corner.approach,
                exitTo: corner.exitTo,
                elevation: corner.elevation,
                camber: corner.camber
            });
        }

        // Step 4: Protocol Assembly (all corners at once)
        if (onProgress) onProgress(80, 'Step 4/4 — Assembling training protocol...');
        const protocol = await this._step4_protocol(processedCorners, template, trackConfig);

        // Merge protocol data into corners
        if (protocol.corners) {
            for (const pc of protocol.corners) {
                const target = processedCorners.find(c => c.number === pc.number);
                if (target) {
                    target.lookAndCall = pc.lookAndCall;
                    target.headRotationCue = pc.headRotationCue;
                    target.speedRamp = pc.speedRamp;
                }
            }
        }

        // Build final blueprint
        const blueprint = {
            id: `bp_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
            trackName: trackConfig.trackName,
            clientName: trackConfig.clientName,
            vehicleType: trackConfig.vehicleType,
            skillLevel: trackConfig.skillLevel,
            generatedAt: new Date().toISOString(),
            templateId: templateId,
            templateVersion: template.version,
            templateChecksum: BlueprintTemplates.getTemplateChecksum(template),
            provider: 'claude_pipeline',
            overallStrategy: protocol.overallStrategy,
            keyPrinciple: protocol.keyPrinciple,
            corners: processedCorners,
            trainingProtocol: protocol.trainingProtocol || {
                dailyMinutes: 15,
                steps: [],
                weakCornerDrills: ''
            }
        };

        if (onProgress) onProgress(100, 'Blueprint complete');
        return blueprint;
    },

    // ── Step 1: Corner Classification ─────────────────────────

    async _step1_classify(corner, template, trackConfig) {
        const cornerJSON = JSON.stringify({
            number: corner.number,
            name: corner.name,
            type: corner.type,
            direction: corner.direction,
            severity: corner.severity,
            approach: corner.approach,
            exitTo: corner.exitTo,
            elevation: corner.elevation,
            camber: corner.camber,
            gazeTargets: corner.gazeTargets,
            kerbDescription: corner.kerbDescription,
            visualChallenge: corner.visualChallenge,
            danVanNotes: corner.danVanNotes
        }, null, 2);

        const prompt = `You are a Quiet Eye corner classification specialist.

TRACK: ${trackConfig.trackName}
VEHICLE: ${template.vehicleType}
SKILL LEVEL: ${trackConfig.skillLevel}

${template.systemPromptSection}

CORNER DATA:
${cornerJSON}

TRACK CONTEXT:
${trackConfig.trackNotes || 'No additional context.'}

CLASSIFY this corner for Quiet Eye protocol design:

RESPOND IN VALID JSON ONLY:
{
  "cornerNumber": ${corner.number || 1},
  "name": "${corner.name || 'Turn'}",
  "type": "hairpin|tight|sweeper|kink|chicane|esses|offcamber|straight|medium",
  "direction": "${corner.direction || 'unknown'}",
  "severity": "very_tight|tight|medium|fast|flat_out",
  "qeDifficulty": "trivial|simple|moderate|challenging|expert",
  "primaryVanTrigger": "target_fixation|early_apex_look|peripheral_threat|instrument_glance|unexpected_threat",
  "gazeDominanceRecommendation": "early_commitment|extended_fixation|rapid_transition",
  "classificationReasoning": "string — 1-2 sentences explaining the QE challenge of this corner"
}`;

        return await this._callWithRetry(prompt, 'classification', template);
    },

    // ── Step 2: Gaze Sequence Generation ──────────────────────

    async _step2_gazeSequence(classified, template, trackConfig) {
        const phaseSpecs = Object.entries(template.gazePhases).map(([phase, spec]) =>
            `${phase.toUpperCase()}:
  Fixation: ${spec.minFixationSec}s–${spec.maxFixationSec}s
  Eyes hint: ${spec.eyesHint}
  Aware hint: ${spec.awareHint}
  DAN fix: ${spec.danFix}`
        ).join('\n\n');

        const prompt = `You are a Quiet Eye gaze sequence engineer.

TRACK: ${trackConfig.trackName}
CORNER: ${classified.name} (${classified.direction} ${classified.type}, QE difficulty: ${classified.qeDifficulty})
VEHICLE: ${template.vehicleType}
PRIMARY VAN TRIGGER: ${classified.primaryVanTrigger}
GAZE DOMINANCE: ${classified.gazeDominanceRecommendation}

${template.systemPromptSection}

VISUAL LANDMARK DATA FOR THIS CORNER:
${classified.gazeTargets ? JSON.stringify(classified.gazeTargets, null, 2) : 'No specific visual data — use corner type defaults.'}
${classified.kerbDescription ? 'KERB: ' + JSON.stringify(classified.kerbDescription, null, 2) : ''}
${classified.visualChallenge ? 'VISUAL CHALLENGE: ' + classified.visualChallenge : ''}
${classified.danVanNotes ? 'DAN/VAN NOTES: ' + classified.danVanNotes : ''}

GAZE PHASE SPECIFICATIONS:
${phaseSpecs}

CRITICAL RULES:
1. "eyes" = SPECIFIC visual feature (kerb colour, board number, landmark). Never generic.
2. "aware" = VEHICLE-SPECIFIC peripheral cue (${template.vehicleType === 'motorcycle' ? 'lean angle, tank edge, handlebar, vestibular' : 'steering wheel, brake pedal, throttle, hand grip'}).
3. Each instruction = ONE clear place for the brain to settle.

RESPOND IN VALID JSON ONLY:
{
  "cornerNumber": ${classified.cornerNumber || classified.number},
  "gazeSequence": {
    "brake": {
      "eyes": "concrete visual target for braking",
      "aware": "vehicle-specific peripheral awareness",
      "fixationDurationSeconds": number
    },
    "apex": {
      "eyes": "concrete visual target at apex",
      "aware": "vehicle-specific peripheral awareness at max load",
      "fixationDurationSeconds": number
    },
    "exit": {
      "eyes": "concrete visual target beyond corner",
      "aware": "vehicle-specific peripheral awareness opening to throttle",
      "fixationDurationSeconds": number
    }
  }
}`;

        return await this._callWithRetry(prompt, 'gazeSequence', template);
    },

    // ── Step 3: Risk Analysis ─────────────────────────────────

    async _step3_riskAnalysis(classified, gazeData, template, trackConfig) {
        const prompt = `You are a Quiet Eye risk analyst specialising in VAN (Ventral Attention Network) suppression.

TRACK: ${trackConfig.trackName}
CORNER: ${classified.name} (${classified.direction} ${classified.type})
VEHICLE: ${template.vehicleType}
QE DIFFICULTY: ${classified.qeDifficulty}
PRIMARY VAN TRIGGER: ${classified.primaryVanTrigger}

GAZE SEQUENCE (just generated for this corner):
${JSON.stringify(gazeData.gazeSequence, null, 2)}

VISUAL CONTEXT:
${classified.visualChallenge || 'No specific visual challenge noted.'}
${classified.danVanNotes || ''}

KNOWN VAN TRIGGERS FOR ${template.vehicleType.toUpperCase()}:
${Object.entries(template.gazePhases).map(([phase, spec]) =>
    `${phase}: ${spec.vanTriggers.join(', ')}`
).join('\n')}

Generate risk analysis and the Quiet Eye settling cue for this corner.

The quietEyeCue must be:
- One powerful sentence the driver/rider says internally
- Contains an ACTION VERB (settle, lock, commit, fixate, snap, hold, anchor)
- References the SPECIFIC visual target from the gaze sequence above
- Ends with vehicle-specific follow-through (e.g. "let the ${template.vehicleType} follow")

RESPOND IN VALID JSON ONLY:
{
  "cornerNumber": ${classified.cornerNumber || classified.number},
  "riskFactors": [
    "specific risk 1",
    "specific risk 2",
    "specific risk 3"
  ],
  "quietEyeCue": "Settle your eyes on [specific target] — let the ${template.vehicleType} follow.",
  "coachingNotes": "2-3 sentences of additional coaching context for this corner"
}`;

        return await this._callWithRetry(prompt, 'riskAnalysis', template);
    },

    // ── Step 4: Training Protocol Assembly ────────────────────

    async _step4_protocol(allCorners, template, trackConfig) {
        const cornerSummary = allCorners.map(c =>
            `Corner ${c.number}: "${c.name}" — ${c.direction} ${c.type} (QE: ${c.qeDifficulty}, VAN: ${c.primaryVanTrigger})`
        ).join('\n');

        const prompt = `You are a Quiet Eye training protocol designer.

TRACK: ${trackConfig.trackName}
VEHICLE: ${template.vehicleType}
SKILL LEVEL: ${trackConfig.skillLevel}
CLIENT: ${trackConfig.clientName || 'Driver'}

${template.systemPromptSection}

ALL CORNERS (classified and with gaze sequences):
${cornerSummary}

SPEED RAMP DEFAULTS FOR ${template.vehicleType.toUpperCase()}:
${JSON.stringify(template.speedRampDefaults, null, 2)}

GENERATE the complete training protocol.

For each corner, provide:
1. lookAndCall — exactly 3 verbal cues (what the driver/rider SAYS aloud during training): "BRAKE!" → "APEX!" → "EXIT!" with specific visual reference
2. headRotationCue — when to begin head rotation (use ${template.vehicleType === 'motorcycle' ? '0.8–1.0s' : '0.55–0.75s'} lead time)
3. speedRamp — "25%"|"50%"|"100%" based on corner difficulty

Also generate:
- overallStrategy: 2-3 sentences on the lap gaze philosophy for this track + vehicle
- keyPrinciple: THE single most important QE insight for this driver
- trainingProtocol: daily training steps (≥5 steps)
- weakCornerDrills: instructions for repeating weak corners

RESPOND IN VALID JSON ONLY:
{
  "overallStrategy": "string",
  "keyPrinciple": "string",
  "corners": [
    {
      "number": 1,
      "lookAndCall": ["BRAKE — eyes on [target]!", "APEX — lock [target]!", "EXIT — snap to [target]!"],
      "headRotationCue": "Begin head rotation X seconds before turn-in...",
      "speedRamp": "25%|50%|100%"
    }
  ],
  "trainingProtocol": {
    "dailyMinutes": 15,
    "steps": [
      { "title": "string", "instruction": "string", "duration": "Xmin" }
    ],
    "weakCornerDrills": "string"
  }
}`;

        return await this._callWithRetry(prompt, 'protocol', template);
    },

    // ── API Caller with Validation & Retry ────────────────────

    async _callWithRetry(prompt, stepName, template, retryCount = 0) {
        const rawText = await AIEngine._callClaudePipeline(prompt);

        let parsed;
        try {
            parsed = BlueprintValidators.extractJSON(rawText);
        } catch (e) {
            if (retryCount < 1) {
                console.warn(`[Pipeline] ${stepName}: JSON parse failed, retrying...`);
                const retryPrompt = prompt + `\n\nIMPORTANT: Your previous response was not valid JSON. Error: ${e.message}\nPlease respond with ONLY valid JSON, no markdown, no explanation.`;
                return this._callWithRetry(retryPrompt, stepName, template, retryCount + 1);
            }
            throw new Error(`Pipeline ${stepName}: Failed to parse JSON after retry. ${e.message}`);
        }

        // Validate based on step
        let validation;
        switch (stepName) {
            case 'classification':
                validation = BlueprintValidators.validateClassification(parsed);
                break;
            case 'gazeSequence':
                validation = BlueprintValidators.validateGazeSequence(parsed, template);
                break;
            case 'riskAnalysis':
                validation = BlueprintValidators.validateRiskAnalysis(parsed, template);
                break;
            case 'protocol':
                validation = BlueprintValidators.validateProtocol(parsed, template);
                break;
            default:
                validation = { valid: true, errors: [], data: parsed };
        }

        if (!validation.valid) {
            if (retryCount < 1) {
                console.warn(`[Pipeline] ${stepName}: Validation failed, retrying...`, validation.errors);
                const errorList = validation.errors.join('\n- ');
                const retryPrompt = prompt + `\n\nIMPORTANT: Your previous response failed validation:\n- ${errorList}\n\nPlease fix these issues and respond with valid JSON only.`;
                return this._callWithRetry(retryPrompt, stepName, template, retryCount + 1);
            }
            console.error(`[Pipeline] ${stepName}: Validation failed after retry:`, validation.errors);
            // Return best-effort data rather than crashing
            return parsed;
        }

        return parsed;
    },

    /**
     * Check if the pipeline is ready to run.
     */
    isConfigured() {
        return !!AIEngine.claudeApiKey && typeof BlueprintTemplates !== 'undefined' && typeof BlueprintValidators !== 'undefined';
    }
};
